README
Xymon-Plugin-Server

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Xymon::Plugin::Server

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Xymon-Plugin-Server

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Xymon-Plugin-Server

    CPAN Ratings
        http://cpanratings.perl.org/d/Xymon-Plugin-Server

    Search CPAN
        http://search.cpan.org/dist/Xymon-Plugin-Server/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Toshimitsu FUJIWARA

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Xymon-Plugin-Server

0.02    26/May/2013
        Makefile.PL
          perl5.6 is not suppoted.
        Xymon::Plugin::Server::Status
          optional hashref parameter is added to constructor.
          ('EscapeMessage' is recognized.)

0.02    11/May/2013
        Xymon::Plugin::Server
          "version" method now returns ARRAYREF correctly.

        Xymon::Plugin::Server::Devmon
          "format" method rename DS name for compatibilty. (for only Xymon 4.2)

0.01    06/May/2013
        First version.


example/test.pl
#!/usr/bin/perl
#
# Xymon::Plugin::Server sample script
#
#
# HOW TO RUN
# ==========
#
# Add "test=devmon" to variable TEST2RRD in xymonserver.cfg.
# (hobbitserver.cfg in Xymon 4.2)
#
#   TEST2RRD="cpu=la,disk,inode,qtree,...,test=devmon"
#
# (After editing this file, restart xymon.)
#
# Add graph definition to graphs.cfg. (hobbitgraph.cfg in Xymon 4.2)
# Definition exists in "add-graph.cfg" in same directory.
#
# Add test name "test" (see Xymon::Plugin::Server::Dispatch constructor
# in this script) at the your host entry in hosts.cfg. (bb-hosts in Xymon 4.2)
#
#   ex.
#     127.0.0.1   localhost.localdomain      # yourtest1 yourtest2 test
#
#  Add etnry to tasks.cfg (hobbitlaunch.cfg in Xymon 4.2) for this script
#  like following:
#
#   [test]
#         ENVFILE $XYMONHOME/etc/hobbitserver.cfg
#         CMD perl /your/path/to/this/script/test.pl
#         LOGFILE $XYMONSERVERLOGS/test.log
#         INTERVAL 5m
#
#   (in Xymon 4.2, "XYMON" is "BB" in variable names)
#
# After all, wait for few minuts...

use strict;

use lib './blib/lib';
use lib '../blib/lib';

use Xymon::Plugin::Server::Dispatch;

package MyMonitor;

use Xymon::Plugin::Server;
use Xymon::Plugin::Server::Status qw(:colors);
use Xymon::Plugin::Server::Devmon;

sub new {
    my $class = shift;
    my ($host, $test, $ip) = @_;

    my $self = {
	host => $host,
	test => $test,
	ip => $ip,
    };

    print "test for $host($ip).$test\n";
    bless $self, $class;
}

sub run {
    my $self = shift;

    my $status = Xymon::Plugin::Server::Status
	->new($self->{host}, $self->{test});

    my $devmon = Xymon::Plugin::Server::Devmon
	->new(ds0 => 'GAUGE:600:0:U',
	      ds1 => 'GAUGE:600:0:U');

    $devmon->add_data(MyData => { ds0 => 0, ds1 => 3 });
    $devmon->add_data(YourData => { ds0 => 8, ds1 => 2 });

    $status->add_status(GREEN, "test1");
    $status->add_status(GREEN, "test2");

    $status->add_message("Helloe world!\nThis is a test message\n");

    $status->add_devmon($devmon);

    $status->add_graph("test");

    $status->report;
}


package main;

my $dispatch = Xymon::Plugin::Server::Dispatch
    ->new('test' => 'MyMonitor');

$dispatch->run;


lib/Xymon/Plugin/Server.pm
package Xymon::Plugin::Server;

use strict;
use warnings FATAL => 'all';

use Carp;

=head1 NAME

Xymon::Plugin::Server - Xymon Server side plugin helper

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';


=head1 DESCRIPTION

Xymon::Plugin::Server package contains some useful modules to write
Xymon plugin executed in server side.

=over

=item  Xymon::Plugin::Server::Status

Status reporter

=item  Xymon::Plugin::Server::Devmon

Data generator in devmon style

=item  Xymon::Plugin::Server::Dispatch

Dispatcher

=back

Check example directory to see how to each modules.


=head1 SUBROUTINES/METHODS

=head2 home

Get Xymon/Hobbit home directory.
This is a class method. call like this:

    my $home = Xymon::Plugin::Server->home;

=cut

sub home {
    return $ENV{XYMONHOME} if (defined($ENV{XYMONHOME})); # xymn 4.3
    return $ENV{BBHOME} if (defined($ENV{BBHOME}));	  # xymn 4.2

    if (defined($ENV{HOME})) {
	my $xyhome = "$ENV{HOME}/server";

	return $xyhome if (-f "$xyhome/etc/hosts.cfg");	# xymon 4.3
	return $xyhome if (-f "$xyhome/etc/bb-hosts");	# xymon 4.2
    }

    croak "environment variable BBHOME/XYMONHOME is not set.";
}

=head2 version

Guess Xymon version and returns resulst as ARRAYREF like [major, minor].

This is a class method. call like this:

    my $ver = Xymon::Plugin::Server->version;

=cut

sub version {
    my $xyhome = __PACKAGE__->home;

    return [4, 3] if (-f "$xyhome/etc/hosts.cfg");	# xymon 4.3

    return [4, 2];
}

=head2 display_hosts

Guess display hosts and returns ARRAY.

This is a class method. call like this:

    my @hosts = Xymon::Plugin::Server->display_hosts;

=cut

sub display_hosts {
    my $display = $ENV{'BBDISP'} || '127.0.0.1';

    return $display if ($display ne '0.0.0.0');

    my $multi_display = $ENV{'BBDISPLAYS'};
    if (defined($multi_display)) {
	return split(/\s+/, $multi_display);
    }

    return $display;
}

=head1 AUTHOR

Toshimitsu FUJIWARA, C<< <tttfjw at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-xymon-plugin-server at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Xymon-Plugin-Server>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Xymon::Plugin::Server


You can also look for information at:

=over 4

=item * About Xymon

L<http://xymon.sourceforge.net/>

=item * About RRDTool

L<http://oss.oetiker.ch/rrdtool/>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Xymon-Plugin-Server>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Xymon-Plugin-Server>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Xymon-Plugin-Server>

=item * Search CPAN

L<http://search.cpan.org/dist/Xymon-Plugin-Server/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Toshimitsu FUJIWARA.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Xymon::Plugin::Server


lib/Xymon/Plugin/Server/Status.pm
#
# status reporter
#
package Xymon::Plugin::Server::Status;

use strict;

use Carp;

use Xymon::Plugin::Server;

=head1 NAME

Xymon::Plugin::Server::Status - Xymon status reporter

=head1 SYNOPSIS

    use Xymon::Plugin::Server::Status qw(:colors);
    my $status = Xymon::Plugin::Server::Status->new("myhostname", "test");

    $status->add_status(GREEN, "this entry is OK");
    $status->add_status(RED, "this entry is NOT OK");
    $status->add_message("Hello! world");

    $status->add_devmon($devmon); # see Xymon::Plugin::Server::Devmon

    $status->add_graph("disk");

    $status->report;  # send status to Xymon server

=head1 EXPORT

Color names

  GREEN YELLOW RED CLEAR PURPLE BLUE

are exported with tag ':colors'

=cut

use base qw(Exporter);

my @colors = qw(GREEN YELLOW RED CLEAR PURPLE BLUE);

our @EXPORT_OK = @colors;
our %EXPORT_TAGS = (colors => \@colors);

use constant {
    GREEN => 'green',
    YELLOW => 'yellow',
    RED => 'red',
    CLEAR => 'clear',
    PURPLE => 'purple',
    BLUE => 'blue',
};

=head1 SUBROUTINES/METHODS

=head2 new(hostname, testname, attr)

Create status object for hostname and testname.

attr is optional hashref to change actions of object.
Currently, following parameter is defined.

=over

=item EscapeMessage

=over

=item 0 (default) 

Make no change to message.

=item 1

Some characters (<, >, &) are replaced to '_'.

=item 2

    '<', '>, '&' are replaced to '&lt;', '&gt;', '&amp' respectively.

=back

=back

=cut

sub new {
    my $class = shift;
    my $host = shift;
    my $test = shift;
    my $attr = shift;

    my $self = {
	_host => $host,
	_test => $test,
	_color => 'clear',
	_message => '',
	_devmon => undef,
	_graph => [],
	_attr => $attr || {}
    };

    bless $self, $class;
}

#
# CLEAR, BLUE, GREEN, PURPLE, YELLOW, RED
#
my %order = (CLEAR	, 0,
	     BLUE	, 1,
	     GREEN	, 2,
	     PURPLE	, 3,
	     YELLOW	, 4,
	     RED	, 5,
	     );

sub _set_color {
    my $self = shift;
    my $color = shift;

    my $cur = $order{$self->{_color}};
    my $new = $order{$color};

    carp "Unknown color: $color" unless (defined($new));

    # if purple is selected, this report will be eliminated by Xymon.
    # so we change color.
    if ($color eq PURPLE) {
	$color = YELLOW;
	$new = $order{$color};
    }
    
    if ($cur < $new) {
	$self->{_color} = $color;
    }

    return $self->{_color};
}

=head2 add_status(color, msg)

Add status and its short message.

=cut

sub add_status {
    my $self = shift;
    my ($color, $msg) = @_;

    if (defined($msg)) {
	$msg .= "\n" if ($msg !~ /\n$/);
	my $m = $self->_escape_string($msg);

	$self->{_message} .= "&$color $m";
    }

    $self->_set_color($color);
}

=head2 add_message(msg)

Add message shown in Xymon status page.

=cut

sub add_message {
    my $self = shift;
    my ($msg) = @_;

    $msg .= "\n" if ($msg !~ /\n$/);

    $self->{_message} .= $self->_escape_string($msg);
}

=head2 add_devmon(devmon)

Add devmon data. See Xymon::Plugin::Server::Devmon

=cut

sub add_devmon {
    my $self = shift;
    my $devmon= shift;

    $self->{_devmon} = $devmon;
}

=head2 add_graph(testname)

Add graph shown in Xymon status page.
"test" name must be defined in graph definition file.
(named hobbitgraph.cfg in Xymon 4.2, graphs.cfg in Xymon 4.3)

=cut

sub add_graph {
    my $self = shift;
    my $test= shift;

    push (@{$self->{_graph}}, $test);
}

sub _create_graph_html {
    my $self = shift;
    my $test = shift;

    my $type = "4.3";

    my ($major, $minor) = @{Xymon::Plugin::Server->version};
    $type = "4.2" if ($major == 4 && $minor == 2);

    my $host = $self->{_host};
    my $color = $self->{_color};
    my $width = $ENV{RRDWIDTH} || 576;
    my $height = $ENV{RRDHEIGHT} || 120;
    my $end_time = time;
    my $start_time = time - (60 * 60 * 48);

    if ($type eq "4.2") {
	my $cgi_url = $ENV{BBSERVERCGIURL} || "/xymon-cgi";

	my $html = << "_EOS";
<p>
<A HREF="$cgi_url/hobbitgraph.sh?host=$host&amp;service=$test&amp;graph_width=$width&amp;graph_height=$height&amp;disp=$host&amp;nostale&amp;color=$color&amp;action=menu">

<IMG BORDER=0 SRC="$cgi_url/hobbitgraph.sh?host=$host&amp;service=$test&amp;graph_width=$width&amp;graph_height=$height&amp;disp=$host&amp;nostale&amp;color=$color&amp;graph=hourly&amp;action=view">
</A>
</p>
_EOS
    }
    else {
	my $cgi_url = $ENV{XYMONSERVERCGIURL} || "/xymon-cgi";

	my $html = << "_EOS";
<p>
<A HREF="$cgi_url/showgraph.sh?host=$host&amp;service=$test&amp;graph_width=$width&amp;graph_height=$height&amp;disp=$host&amp;nostale&amp;color=$color&amp;graph_start=$start_time&amp;graph_end=$end_time&amp;action=menu">
<IMG BORDER=0 SRC="$cgi_url/showgraph.sh?host=$host&amp;service=$test&amp;graph_width=$width&amp;graph_height=$height&amp;disp=$host&amp;nostale&amp;color=$color&amp;graph_start=$start_time&amp;graph_end=$end_time&amp;graph=hourly&amp;action=view">
</A>
</p>
_EOS

    }
}

sub _escape_replace {
    my $s = shift;
    $s =~ s/[<>&]/_/g;
    return $s;
}

sub _escape_entity {
    my $s = shift;

    $s =~ s/&/&amp;/g;
    $s =~ s/</&lt;/g;
    $s =~ s/>/&gt;/g;
    return $s;
}

sub _escape_string {
    my $self = shift;
    my $s = shift;

    if ($self->{_attr}->{EscapeMessage}) {
	my $n = $self->{_attr}->{EscapeMessage};
	if ($n == 1) {
	    return _escape_replace($s);
	}
	elsif ($n == 2) {
	    return _escape_entity($s);
	}

    }

    return $s;
}

sub _create_report_msg {
    my $self = shift;

    local $ENV{LANG} = 'C';
    my $datestr = scalar localtime time;
    my $statstr = '';

    if ($self->{_color} eq GREEN) {
	$statstr = ' - OK';
    }
    elsif ($self->{_color} eq YELLOW || $self->{_color} eq RED) {
	$statstr = ' - NOT OK';
    }

    my $msg = sprintf("status %s.%s %s %s %s%s\n",
		      $self->{_host},
		      $self->{_test},
		      $self->{_color},
		      $datestr,
		      $self->{_test},
		      $statstr);

    $msg .= "\n";
    $msg .= $self->{_message};

    if ($self->{_devmon}) {
	$msg .= "\n";
	$msg .= $self->{_devmon}->format;
    }

    if (@{$self->{_graph}} > 0) {
	$msg .= "</PRE></TD></TR></TABLE>\n";

	for my $g (@{$self->{_graph}}) {
	    $msg .= $self->_create_graph_html($g);
	}

	$msg .= "<TABLE><TR><TD><PRE>\n";
    }

    return $msg;
}

=head2 report

Report current status to Xymon server.

=cut

sub report {
    my $self = shift;

    my $msg = $self->_create_report_msg;

    print "$msg" if ($ENV{XYMON_PLUGIN_DEBUG});

    my $bb = Xymon::Plugin::Server->home . "/bin/bb";

    for my $bbh (Xymon::Plugin::Server->display_hosts) {
	print "send to $bbh\n" if ($ENV{XYMON_PLUGIN_DEBUG});
	open(my $fh, "|-", "$bb", $bbh, "@") or die "cannot execute $bb: $!";
	print $fh $msg;
    }
}

1;


lib/Xymon/Plugin/Server/Dispatch.pm
#
# plugin dispatcher
#
package Xymon::Plugin::Server::Dispatch;

use strict;
use warnings;

=head1 NAME

Xymon::Plugin::Server::Dispatch - Xymon plugin dispatcher

=head1 SYNOPSIS

    use Xymon::Plugin::Server::Dispatch;
    use YourMonitor;

    # dispatch to class
    my $dispatch1 = Xymon::Plugin::Server::Dispatch
                    ->new('test' => 'YourMonitor');
    $dispatch1->run;

    # dispatch to method
    my $dispatch2 = Xymon::Plugin::Server::Dispatch
                    ->new('test' => new YourMonitor());
    $dispatch2->run;

    # dispatch to CODEREF
    my $dispatch3 = Xymon::Plugin::Server::Dispatch
                    ->new('test' => sub { ... });
    $dispatch3->run;


=cut;

use Xymon::Plugin::Server::Hosts;


=head1 SUBROUTINES/METHODS

=head2 new(testName1 => ModuleName1, ...)

Create dispatch object for tests and modules.

If testName has wildcard character(like http:*), $test will be ARRAYREF
when run method is called.

=cut

sub new {
    my $class = shift;
    my @keyvalue = @_;

    my @keys;
    my %dic;

    while (@keyvalue > 0) {
	my $key = shift @keyvalue;
	push(@keys, $key);
	$dic{$key} = shift @keyvalue;
    }

    my $self = {
	_keys => \@keys,
	_dic => \%dic,
    };

    bless $self, $class;
}

=head1 SUBROUTINES/METHODS

=head2 run

For every host listed in bb-hosts(Xymon 4.2) or hosts.cfg (Xymon 4.3),
following operation is executed.

    # if class name is given
    my $module = YourMonitor->new($host, $test);
    $module->run;

    # if object is given
    $module->run($host, $test);

    # if CODEREF is given
    &$code($host, $test);

=cut

sub run {
    my $self = shift;

    for my $key (@{$self->{_keys}}) {
	my $dest = $self->{_dic}->{$key};
	my $code;
	if (ref($dest)) {
	    if (ref($dest) eq 'CODE') {
		$code = $dest;
	    }
	    else {
		$code = sub { $dest->run(@_); };
	    }
	}
	else {
	    $code = sub {
		my $obj = $dest->new(@_);
		$obj->run;
	    };
	}

	for my $entry (Xymon::Plugin::Server::Hosts->new->grep($key)) {
	    eval {
		my $host = $entry->[1];
		my $test = $entry->[2];
		my $ip = $entry->[0];

		&$code($host, $test, $ip);
	    };
	    if ($@) {
		print STDERR $@;
	    }
	}
    }
}

1;


lib/Xymon/Plugin/Server/Devmon.pm
#
# Devmon data object
#
package Xymon::Plugin::Server::Devmon;

use strict;

=head1 NAME

Xymon::Plugin::Server::Devmon - Devmon data object

=head1 SYNOPSIS

    my $devmon = Xymon::Plugin::Server::Devmon
        ->new(ds0 => 'GAUGE:600:0:U',
              ds1 => 'GAUGE:600:0:U');

    $devmon->add_data(device1 => { ds0 => 0, ds1 => 2 });
    $devmon->add_data(device2 => { ds0 => 0, ds1 => 2 });

    # add to data to be reported.
    my $status = Xymon::Plugin::Server::Status
                 ->new("localhost.localdomain", "test1");

    $status->add_devmon($devmon);

=head1 DESCRIPTION

This module handles data structure for Xymon 'devmon' module.
To store data into RRD database, add entry to TEST2RRD variable
 in server config.

(named hobbitserver.cfg in Xymon 4.2, xymonserver.cfg in Xymon 4.3)

 ex.
  TEST2RRD="cpu=la,disk,...(snipped)...,test1=devmon"


=head1 SUBROUTINES/METHODS

=head2 new(ds1 => dsdef1, ...)

Create devmon object with data store definition.
(To know meanings of dsdef field, please read document in RRDTool.)

=cut

sub new {
    my $class = shift;
    my @datadef = @_;	# key, def, key, def, ...

    my @dskeys;
    my %dsdefs;

    while (@datadef) {
	my $key = shift @datadef;
	push(@dskeys, $key);
	$dsdefs{$key} = shift @datadef;
    }

    my $self = {
	_dskeys => \@dskeys,
	_dsdefs => \%dsdefs,
	_data => {},
    };

    bless $self, $class;
}

=head2 add_data(devname, { ds1 => ds1value, ...})

Set values for devname.

=cut

sub add_data {
    my $self = shift;
    my ($key, $values) = @_;

    $self->{_data}->{$key} = $values;
}

=head2 format

Format data structure to report to Xymon.

=cut

sub format {
    my $self = shift;
    my @ret;

    push(@ret, "<!--DEVMON RRD: ");
    my $dsline = join(" ",
		      map { join(':', 'DS', $_, $self->{_dsdefs}->{$_}) }
		      @{$self->{_dskeys}});

    push(@ret, $dsline);

    while (my ($key, $values) = each %{$self->{_data}}) {
	
	my $line = $key . " "
	    . join(":",
		   map { defined($values->{$_}) ? $values->{$_} : ''; }
		   @{$self->{_dskeys}});

	push(@ret, $line);
    }

    push(@ret, "-->\n");

    return join("\n", @ret); 
}

1;
